ord_schema/interface/query.py

```diff
@@ -20,15 +47,252 @@
+            SELECT DISTINCT reaction_id, serialized 
+            FROM reactions 
+            WHERE reaction_id = ANY (%s)""")
+        args = [self._reaction_ids]
+        logging.info('Running SQL command:%s',
+                     cursor.mogrify(query, args).decode())
+        cursor.execute(query, args)
+        return fetch_results(cursor)
+
+
+class ReactionSmartsQuery(ReactionQueryBase):
+    """Matches reactions by reaction SMARTS."""
+
+    def __init__(self, reaction_smarts):
+        """Initializes the query.
+
+        Args:
+            reaction_smarts: Reaction SMARTS.
+        """
+        self._reaction_smarts = reaction_smarts
+
+    def json(self):
+        """Returns a JSON representation of the query."""
+        return json.dumps({'reactionSmarts': self._reaction_smarts})
+
+    def run(self, cursor, limit=None):
+        """Runs the query.
+
+        Args:
+            cursor: psycopg.cursor instance.
+            limit: Integer maximum number of matches. If None (the default), no
+                limit is set.
+
+        Returns:
+            Dict mapping reaction IDs to serialized Reaction protos.
+        """
+        components = [
+            sql.SQL("""
+            SELECT DISTINCT reaction_id, serialized
+            FROM reactions
+            INNER JOIN rdk.reactions USING (reaction_id)
+            WHERE rdk.reactions.r@>reaction_from_smarts(%s)""")
+        ]
+        args = [self._reaction_smarts]
+        if limit:
+            components.append(sql.SQL(' LIMIT %s'))
+            args.append(limit)
+        query = sql.Composed(components).join('')
+        logging.info('Running SQL command:%s',
+                     cursor.mogrify(query, args).decode())
+        cursor.execute(query, args)
+        return fetch_results(cursor)
+
+
+class ReactionComponentQuery(ReactionQueryBase):
+    """Matches reactions by reaction component predicates."""
+
+    def __init__(self, predicates, do_chiral_sss=False, tanimoto_threshold=0.5):
+        """Initializes the query.
+
+        Args:
+            predicates: List of ReactionComponentPredicate objects.
+            do_chiral_sss: If True, consider stereochemistry in substructure
+                searches.
+            tanimoto_threshold: Float Tanimoto similarity threshold. Pairs
+                below this threshold will not be considered matches.
+        """
+        self._predicates = predicates
+        self._do_chiral_sss = do_chiral_sss
+        self._tanimoto_threshold = tanimoto_threshold
+
+    def json(self):
+        """Returns a JSON representation of the query."""
+        return json.dumps({
+            'useStereochemistry': self._do_chiral_sss,
+            'similarity': self._tanimoto_threshold,
+            'components': [predicate.json() for predicate in self._predicates],
+        })
 
-class Predicate:
-    """Structure and code generation for ORD query predicates."""
+    def _setup(self, cursor):
+        """Prepares the database for a query.
```

Comment:
This is fine, but I thought things were more readable when all the SQL was a simple string. Is there an advantage in going command/args/logging/execute for each statement? Maybe something about sql.SQL?

Reply:
The psycopg2 documentation is very clear about this:

> __Warning__ Never, __never__, __NEVER__ use Python string concatenation (+) or string parameters interpolation (%) to pass variables to a SQL query string. Not even at gunpoint.

https://www.psycopg.org/docs/usage.html#the-problem-with-the-query-parameters

The `sql` module is the preferred way to construct queries where you need to use variables for SQL identifiers like table names. Simple queries that have static identifiers should use the `%s` syntax and pass `args` to `execute`.